### PR TITLE
session: add gofail test for (#8743)

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Add gofail test for https://github.com/pingcap/tidb/pull/8743

### What is changed and how it works?
Add a gofail test.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8771)

### CODE REVIEW

**Comment by shenli** [comment]
Good job!

**Comment by shenli** [comment]
@crazycs520 Please fill the PR template.

File: session/txn.go [hunk 1]
```diff
@@ -277,11 +277,22 @@ type txnFuture struct {
...
+
 func (tf *txnFuture) wait() (kv.Transaction, error) {
 	if tf.mockFail {
 		return nil, errors.New("mock get timestamp fail")
 	}
 
+	// mockGetTSErrorInRetry should wait mockCommitErrorOnce first, then will run into retry() logic.
+	// Then mockGetTSErrorInRetry will return retryable error when first retry.
+	// Before PR #8743, we don't cleanup txn after meet error such as error like: PD server timeout[try again later]
+	// This may cause multiple identical data to be written.
```
Comment (shenli) [inline session/txn.go]: multiple identical -> duplicate

File: session/session.go [hunk 2]
```diff
@@ -282,6 +282,8 @@ func (s *schemaLeaseChecker) Check(txnTS uint64) error {
 	return domain.ErrInfoSchemaExpired
 }
 
+var mockCommitErrorOnce = true
```
Comment (shenli) [inline session/session.go]: Add a comment for this variable.

Comment (crazycs520) [reply session/session.go]: done.

File: session/txn.go (hunk 1 above)
Comment (crazycs520) [reply session/txn.go]: done.

**Comment by zimulala** [comment]
/run-all-tests tidb-test=release-2.0 tikv=release-2.0 pd=release-2.0

**Comment by zz-jason** [comment]
@crazycs520 Do we have a corresponding PR in the master branch?

**Pull request merged**